### PR TITLE
Relations via array

### PR DESCRIPTION
**Pull request by Tigrov**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #362

### CODE REVIEW

File: src/ActiveRelationTrait.php [hunk 1]
```diff
@@ -12,6 +12,11 @@
 use Yiisoft\Db\Exception\InvalidConfigException;
 use Yiisoft\Db\Exception\NotSupportedException;
```
Comment (vjik) [inline src/ActiveRelationTrait.php:14]:
```suggestion
```

File: src/ActiveRelationTrait.php [hunk 2]
```diff
@@ -533,31 +538,45 @@ protected function filterByModels(array $models): void
...
-                    if (!empty($value)) {
-                        $values[] = array_combine($attributes, array_merge($nulls, $value));
-                    }
+            $scalarValues = array_unique($scalarValues);
+            $values = [...$scalarValues, ...$nonScalarValues];
+
+            $attribute = reset($attributes);
+
+            match ($this->getARInstance()->columnType($attribute)) {
+                'array' => $this->andWhere(new ArrayOverlapsCondition($attribute, $values)),
```
Comment (vjik) [inline src/ActiveRelationTrait.php:556]: Should we add constant `TYPE_ARRAY` to `SchemaInterface` in Yii DB?

File: src/ActiveRelationTrait.php [hunk 3]
```diff
@@ -533,31 +538,45 @@ protected function filterByModels(array $models): void
...
+            }
+        } else {
+            foreach ($models as $model) {
+                $value = array_intersect_key($model, $nulls);
 
-                    if (!empty($value)) {
-                        $values[] = array_combine($attributes, array_merge($nulls, $value));
-                    }
+                if (!empty($value)) {
+                    $values[] = array_combine($attributes, array_merge($nulls, $value));
```
Comment (samdark) [inline src/ActiveRelationTrait.php:579]: Could be optimized: https://www.exakat.io/en/speeding-up-array_merge/

File: src/ActiveRelationTrait.php (hunk 2 above)
Comment (Tigrov) [reply src/ActiveRelationTrait.php:556]: Yes, plan to move `TYPE_ARRAY` and `TYPE_BIT` from `db-pgsql` to `db` package

File: src/ActiveRelationTrait.php (hunk 3 above)
Comment (Tigrov) [reply src/ActiveRelationTrait.php:579]: This is a bit different case. `array_merge($nulls, $value)` prepares values and then `array_combine()` combines them with prepared keys `$attributes`.

**Pull request merged**